hack/version2.py

```diff
@@ -0,0 +1,238 @@
+                level="fatal",
+            )
+
+        # Next look for named descendants of our commit. This indicates we're versioning an old commit on the branch.
+        if len(descendants) == 1:
+            branch = list(descendants)[0]
+            log(
+                f"Found branch {branch} descended from commit {self.shortsha}",
+                level="debug",
+            )
+            return branch
+
+        log(f"descendants: {descendants}", level="debug")
+
+        if len(descendants) == 0:
+            log(f"WARNING: Are you versioning an unmerged commit?", level="warning")
+
+        if len(ancestors) == 1:
+            branch = list(ancestors)[0]
+            log(
+                f"Found branch {branch} which is an ancestor of commit {self.shortsha}",
+                level="debug",
+            )
+            return branch
+
+        log(
+            f"Could not determine a branch! Please specify one explicitly.",
+            level="fatal",
+        )
+
+    @property
+    def shortsha(self):
+        return self.commit.hexsha[0:7]
+
+    def _prefix_from_branch(self):
+        """Determine the X.Y semver prefix for the branch name.
+
+        :return prefix (str): Two-digit semver prefix corresponding to the branch.
+            - If the branch name is of the form [remote/]mce-X.Y, this will be X.Y.
+            - If the branch name is [remote/]master, we'll use the master prefix.
+            - Otherwise, we'll use the "unknown" default.
+        """
+        m = MCE_BRANCH_RE.match(self._branch_name)
+        if m:
+            return m.group(2)
+
+        m = MASTER_BRANCH_RE.match(self._branch_name)
+        if m:
+            return MASTER_BRANCH_PREFIX
+
+        return UNKNOWN_BRANCH_PREFIX
+
+    def _find_branch(self, branch_name: str):
+        # The user may have specified a branch name for which we don't have a local ref
+        # (e.g. `mce-2.1`). Rather than forcing them to specify a qualified ref, we'll
+        # automatically look for it in the `upstream` and `origin` remotes as well.
+        for remote in ("", "upstream/", "origin/"):
+            branch = remote + branch_name
+            log(f"Trying to find {branch_name} at {branch}")
+            try:
+                commit = self.repo.rev_parse(branch)
+            except git.BadName:
+                log(f"Couldn't find {branch_name} at `{branch}`")
+            else:
+                log(f"Found {branch_name} at `{branch}`")
+                break
+        else:
+            log(f"Couldn't find {branch_name} branch!", level="fatal")
+        return commit
+
+    @property
+    def semver(self):
+        return f"{self._prefix}.{self._commit_count}-{self.shortsha}"
+
+    def __str__(self) -> str:
+        return f"v{self.semver}"
+
+
+def log(*args, level="info", **kwargs):
+    if _mode == "standalone":
+        print(*args, file=sys.stderr, **kwargs)
+    else:
+        print(*args, **kwargs)
+    if level == "fatal":
```

Comment:
I've been seeing this pattern in go code and kinda like it. It does a couple of things:
- One line instead of two
- If we decide to change what a "fatal" error does -- like, who knows, maybe it always writes a stack trace? -- we only have to change one place.